File: clients/javascript/openmemory-client.js
/**
 * OpenMemory JavaScript/TypeScript Client
 * 
 * A simple HTTP client for the OpenMemory ICP canister.
 */

class OpenMemoryClient {
    constructor(baseUrl, authToken) {
        this.baseUrl = baseUrl.replace(/\/$/, '');
        this.authToken = authToken;
    }

    async request(endpoint, options = {}) {
        const url = `${this.baseUrl}${endpoint}`;
        const headers = {
            'Content-Type': 'application/json',
            ...options.headers,
        };

        if (this.authToken) {
            headers['Authorization'] = `Bearer ${this.authToken}`;
        }

        const response = await fetch(url, { ...options, headers });
        
        if (!response.ok) {
            const error = await response.json().catch(() => ({ error: 'Unknown error' }));
            throw new Error(`HTTP ${response.status}: ${error.error || error.message}`);
        }

        return response;
    }

    async addMemory(content, metadata = {}, tags = []) {
        const response = await this.request('/memories', {
            method: 'POST',
            body: JSON.stringify({
                content,
                metadata,
                tags,
            }),
        });

        const result = await response.json();
        return result.id;
    }

    async searchMemories(query, options = {}) {
        const params = new URLSearchParams({ q: query });
        if (options.limit) params.set('limit', options.limit.toString());
        if (options.tags) params.set('tags', options.tags);

        const response = await this.request(`/memories/search?${params}`);
        const result = await response.json();
        return result.results;
    }

    async getMemory(id) {
        const response = await this.request(`/memories/${id}`);
        return response.json();
    }

    async deleteMemory(id) {
        const response = await this.request(`/memories/${id}`, { method: 'DELETE' });
        const result = await response.json();
        return result.deleted || false;
    }

    async listMemories(offset = 0, limit = 20) {
        const params = new URLSearchParams({ 
            offset: offset.toString(), 
            limit: limit.toString() 
        });
        
        const response = await this.request(`/memories?${params}`);
        const result = await response.json();
        return result.memories;
    }

    async healthCheck() {
        const response = await this.request('/health');
        return response.json();
    }

    async getStats() {
        const response = await this.request('/stats');
        return response.json();
    }
}

// Example usage
async function example() {
    const client = new OpenMemoryClient(
        'https://rdmx6-jaaaa-aaaaa-aaadq-cai.ic0.app',
        'test-token'
    );

    try {
        // Health check
        const health = await client.healthCheck();
        console.log(`Service status: ${health.status}`);

        // Add memory
        const memoryId = await client.addMemory(
            "Rustは安全で高速なシステムプログラミング言語です",
            { source: "documentation", language: "ja" },
            ["Rust", "programming", "systems"]
        );
        console.log(`Created memory: ${memoryId}`);

        // Search
        const memories = await client.searchMemories("Rustの特徴", { limit: 5 });
        console.log(`Found ${memories.length} memories`);
        memories.forEach(result => {
            const preview = result.memory.content.length > 50 
                ? result.memory.content.substring(0, 50) + "..."
                : result.memory.content;
            console.log(`  - ${preview} (score: ${result.similarity_score.toFixed(2)})`);
        });

        // Statistics
        const stats = await client.getStats();
        console.log(`Service stats:`, stats);

    } catch (error) {
        console.error('Error:', error.message);
    }
}

// Node.js module export
if (typeof module !== 'undefined' && module.exports) {
    module.exports = OpenMemoryClient;
}

// Browser global
if (typeof window !== 'undefined') {
    window.OpenMemoryClient = OpenMemoryClient;
}

// Example for Node.js
if (typeof require !== 'undefined' && require.main === module) {
    example();
}

File: clients/python/openmemory_client.py
#!/usr/bin/env python3
"""
OpenMemory Python Client

A simple HTTP client for the OpenMemory ICP canister.
"""

import requests
from typing import List, Dict, Optional
import json


class OpenMemoryClient:
    def __init__(self, base_url: str, auth_token: Optional[str] = None):
        """
        Initialize the OpenMemory client.
        
        Args:
            base_url: The base URL of the OpenMemory canister
            auth_token: Optional authentication token
        """
        self.base_url = base_url.rstrip('/')
        self.session = requests.Session()
        if auth_token:
            self.session.headers.update({'Authorization': f'Bearer {auth_token}'})
    
    def add_memory(self, content: str, metadata: Dict = None, tags: List[str] = None) -> str:
        """
        Add a new memory.
        
        Args:
            content: The memory content
            metadata: Optional metadata dictionary
            tags: Optional list of tags
            
        Returns:
            The memory ID
        """
        payload = {
            'content': content,
            'metadata': metadata or {},
            'tags': tags or []
        }
        
        response = self.session.post(f'{self.base_url}/memories', json=payload)
        response.raise_for_status()
        return response.json()['id']
    
    def search_memories(self, query: str, limit: int = 10, tags: str = None) -> List[Dict]:
        """
        Search memories using semantic search.
        
        Args:
            query: Search query
            limit: Maximum number of results
            tags: Comma-separated list of tags to filter by
            
        Returns:
            List of search results
        """
        params = {'q': query, 'limit': limit}
        if tags:
            params['tags'] = tags
            
        response = self.session.get(f'{self.base_url}/memories/search', params=params)
        response.raise_for_status()
        return response.json()['results']
    
    def get_memory(self, memory_id: str) -> Dict:
        """
        Get a specific memory by ID.
        
        Args:
            memory_id: The memory ID
            
        Returns:
            The memory object
        """
        response = self.session.get(f'{self.base_url}/memories/{memory_id}')
        response.raise_for_status()
        return response.json()
    
    def delete_memory(self, memory_id: str) -> bool:
        """
        Delete a memory.
        
        Args:
            memory_id: The memory ID
            
        Returns:
            True if deleted successfully
        """
        response = self.session.delete(f'{self.base_url}/memories/{memory_id}')
        response.raise_for_status()
        return response.json().get('deleted', False)
    
    def list_memories(self, offset: int = 0, limit: int = 20) -> List[Dict]:
        """
        List memories with pagination.
        
        Args:
            offset: Number of memories to skip
            limit: Maximum number of memories to return
            
        Returns:
            List of memories
        """
        params = {'offset': offset, 'limit': limit}
        response = self.session.get(f'{self.base_url}/memories', params=params)
        response.raise_for_status()
        return response.json()['memories']
    
    def health_check(self) -> Dict:
        """
        Check the health status of the service.
        
        Returns:
            Health status information
        """
        response = self.session.get(f'{self.base_url}/health')
        response.raise_for_status()
        return response.json()
    
    def get_stats(self) -> Dict:
        """
        Get service statistics.
        
        Returns:
            Statistics about the service
        """
        response = self.session.get(f'{self.base_url}/stats')
        response.raise_for_status()
        return response.json()


def main():
    """Example usage of the OpenMemory client."""
    
    # Initialize client (replace with your canister URL)
    client = OpenMemoryClient(
        'https://rdmx6-jaaaa-aaaaa-aaadq-cai.ic0.app',
        auth_token='test-token'  # Use appropriate token
    )
    
    try:
        # Health check
        health = client.health_check()
        print(f"Service status: {health['status']}")
        
        # Add a memory
        memory_id = client.add_memory(
            content="ICPは分散型クラウドコンピュータプラットフォームです",
            metadata={"source": "documentation", "language": "ja"},
            tags=["ICP", "blockchain", "tech"]
        )
        print(f"Created memory: {memory_id}")
        
        # Search for memories
        results = client.search_memories("ICPの特徴", limit=5)
        print(f"Found {len(results)} relevant memories:")
        for result in results:
            content_preview = result['memory']['content'][:50] + "..." if len(result['memory']['content']) > 50 else result['memory']['content']
            print(f"  - {content_preview} (score: {result['similarity_score']:.2f})")
        
        # List all memories
        memories = client.list_memories(limit=10)
        print(f"Total memories: {len(memories)}")
        
        # Get statistics
        stats = client.get_stats()
        print(f"Service stats: {stats}")
        
    except requests.RequestException as e:
        print(f"Error: {e}")


if __name__ == "__main__":
    main()